datafusion/core/src/physical_optimizer/join_pipeline_selection.rs

```diff
@@ -703,7 +898,266 @@ fn select_best_streaming_plan(
 ) -> Option<Arc<dyn ExecutionPlan>> {
     possible_plans
         .into_iter()
-        .find(|plan| is_plan_streaming(plan))
+        .find(|plan| is_plan_streaming_with_error(plan).is_ok())
+}
+
+/// Determines if the given execution plan node type is allowed before encountering
+/// a `PartitionedHashJoinExec` node in the plan tree.
+///
+/// # Parameters
+/// * `plan`: The execution plan node to check.
+///
+/// # Returns
+/// * Returns `true` if the node is one of the following types:
+///   * `CoalesceBatchesExec`
+///   * `SortExec`
+///   * `CoalescePartitionsExec`
+///   * `SortPreservingMergeExec`
+///   * `ProjectionExec`
+///   * `RepartitionExec`
+///   * `PartitionedHashJoinExec`
+/// * Otherwise, returns `false`.
+///
+fn is_allowed(plan: &Arc<dyn ExecutionPlan>) -> bool {
+    let as_any = plan.as_any();
+    as_any.is::<CoalesceBatchesExec>()
+        || as_any.is::<SortExec>()
+        || as_any.is::<CoalescePartitionsExec>()
+        || as_any.is::<SortPreservingMergeExec>()
+        || as_any.is::<ProjectionExec>()
+        || as_any.is::<RepartitionExec>()
+        || as_any.is::<PartitionedHashJoinExec>()
+}
+
+/// Recursively checks the execution plan from the given node downward to determine
+/// if any unallowed executors exist before a `PartitionedHashJoinExec` node.
```

Comment:
```suggestion
/// if any unallowed executors exist after a `PartitionedHashJoinExec` node.
```
I am not sure though, can you double check